src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | Errors that maybe raised from the ledgers methods
+data LedgerError c
+    = StateNotInCache (BlockHash c)
+    -- ^ State for the given block is not materialized in the state
+    -- cache.
+
+deriving instance (Crypto.HasHashing c) => Eq (LedgerError c)
+deriving instance (Show (Crypto.Hash c)) => Show (LedgerError c)
+
+data Ledger crypto seal tx output state (m :: * -> *) = Ledger
+    { blockStore   :: BlockStore crypto tx seal m
+    , stateStore   :: HashStore crypto state m
+    , receiptStore :: ReceiptStore.ReceiptStore crypto tx output m
+    , ledgerEval   :: Evaluator state tx output
+    }
+
+hoist
+    :: (forall a. m a -> n a)
+    -> Ledger crypto seal tx output state m
+    -> Ledger crypto seal tx output state n
+hoist f l =
+    Ledger
+        { blockStore = BlockStore.hoistBlockStore f $ blockStore l
+        , stateStore = hoistHashStore f $ stateStore l
+        , receiptStore = ReceiptStore.hoistReceiptStore f $ receiptStore l
+        , ledgerEval = ledgerEval l
+        }
+
+
+-- | Create a ledger using in-memory storage for blocks, states, and
+-- receipts.
+--
+-- Uses 'newFromBlockStore' with 'newBlockStoreIO' initialized with the
+-- given blocks.
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newMemoryIO
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> Chrono.NewestFirst NonEmpty (Block crypto tx (Sealed crypto seal))
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryIO ledgerEval blks initialState = do
+    blockStore <- newBlockStoreIO blks
+    newFromBlockStore ledgerEval blockStore initialState
+
+
+-- | Same as 'newMemoryIO' with just @genesisBlk@ in the blockchain.
+newMemoryFromGenesisIO
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> Block crypto tx (Sealed crypto seal)
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryFromGenesisIO ledgerEval genesisBlk initialState =
+    let blks = Chrono.NewestFirst $ genesisBlk NonEmpty.:| []
+    in newMemoryIO ledgerEval blks initialState
+
+
+-- | Create a ledger backed by the given block store and in-memory
+-- storage of state and receipts.
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newFromBlockStore
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> BlockStore crypto tx seal IO
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newFromBlockStore ledgerEval blockStore initialState = do
+    receiptStore <- ReceiptStore.newReceiptStoreIO
+    stateStore <- newHashStoreIO
+    storeHashContent stateStore initialState
```

Comment:
I know this isn't part of this PR, but this function reads funny, as "storing the contents of a hash" :smile: